carstore/bs.go

```diff
@@ -930,49 +931,320 @@ func (cs *CarStore) Stat(ctx context.Context, usr models.Uid) ([]UserStat, error
 	return out, nil
 }
 
-func (cs *CarStore) checkFork(ctx context.Context, user models.Uid, prev cid.Cid) (bool, error) {
-	ctx, span := otel.Tracer("carstore").Start(ctx, "checkFork")
-	defer span.End()
+func (cs *CarStore) TakeDownRepo(ctx context.Context, user models.Uid) error {
+	var shards []CarShard
+	if err := cs.meta.Find(&shards, "usr = ?", user).Error; err != nil {
+		return err
+	}
 
-	lastShard, err := cs.getLastShard(ctx, user)
+	for _, sh := range shards {
+		if err := cs.deleteShard(ctx, &sh); err != nil {
+			if !os.IsNotExist(err) {
+				return err
+			}
+		}
+	}
+
+	if err := cs.meta.Delete(&CarShard{}, "usr = ?", user).Error; err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (cs *CarStore) deleteShard(ctx context.Context, sh *CarShard) error {
+	if err := cs.meta.Delete(&CarShard{}, "id = ?", sh.ID).Error; err != nil {
+		return err
+	}
+
+	if err := cs.meta.Delete(&blockRef{}, "shard = ?", sh.ID).Error; err != nil {
+		return err
+	}
+
+	return cs.deleteShardFile(ctx, sh)
+}
+
+type shardStat struct {
+	ID    uint
+	Seq   int
+	Dirty int
+	Total int
+
+	refs []blockRef
+}
+
+func (s shardStat) dirtyFrac() float64 {
+	return float64(s.Dirty) / float64(s.Total)
+}
+
+func shouldCompact(s shardStat) bool {
+	// if shard is mostly removed blocks
+	if s.dirtyFrac() > 0.5 {
+		return true
+	}
+
+	// if its a big shard with a sufficient number of removed blocks
+	if s.Dirty > 1000 {
+		return true
+	}
+
+	// if its just rather small and we want to compact it up with other shards
+	if s.Total < 20 {
+		return true
+	}
+
+	return false
+}
+
+func aggrRefs(brefs []blockRef) []shardStat {
+	byId := make(map[uint]*shardStat)
+
+	for _, br := range brefs {
+		s, ok := byId[br.Shard]
+		if !ok {
+			s = &shardStat{
+				ID: br.Shard,
+			}
+			byId[br.Shard] = s
+		}
+
+		s.Total++
+		if br.Dirty {
+			s.Dirty++
+		}
+
+		s.refs = append(s.refs, br)
+	}
+
+	var out []shardStat
+	for _, s := range byId {
+		out = append(out, *s)
+	}
+
+	sort.Slice(out, func(i, j int) bool {
+		return out[i].ID < out[j].ID
+	})
+
+	return out
+}
+
+type compBucket struct {
+	shards []shardStat
+
+	cleanBlocks int
+}
+
+func (cb *compBucket) addShardStat(ss shardStat) {
+	cb.cleanBlocks += (ss.Total - ss.Dirty)
+	cb.shards = append(cb.shards, ss)
+}
+
+func (cb *compBucket) isFull() bool {
+	return cb.cleanBlocks > 50
```

Comment:
Is this a magic number? If so can it have a comment explaining why it was chosen?